src/platforms/rp2040/src/lib/gpiodriver.c

```diff
@@ -117,23 +128,44 @@ static term nif_gpio_digital_write(Context *ctx, int argc, term argv[])
 {
     UNUSED(argc);
 
-    VALIDATE_VALUE(argv[0], term_is_integer);
-    int gpio_num = term_to_int(argv[0]);
+    term gpio_pin = argv[0];
     term level_term = argv[1];
 
     int level;
     if (term_is_integer(level_term)) {
         level = term_to_int32(level_term);
         if (UNLIKELY((level != 0) && (level != 1))) {
-            return ERROR_ATOM;
+            RAISE_ERROR(BADARG_ATOM);
         }
     } else {
         level = interop_atom_term_select_int(pin_level_table, level_term, ctx->global);
         if (UNLIKELY(level < 0)) {
-            return ERROR_ATOM;
+            RAISE_ERROR(BADARG_ATOM);
         }
     }
-    gpio_put(gpio_num, level);
+
+    uint gpio_num;
+    if (term_is_integer(gpio_pin)) {
+        gpio_num = gpio_pin;
```

Comment:
Oops. gpio_pin is a term and gpio_num is an int, so we need some variant of term_to_int.

Also I wonder if we should friendly check the gpio number range and raise badarg if needed as the underlying API doesn't at all unless assertions are enabled. The constant is NUM_BANK0_GPIOS.

https://github.com/raspberrypi/pico-sdk/blob/6a7db34ff63345a7badec79ebea3aaef1712f374/src/rp2_common/hardware_gpio/include/hardware/gpio.h#L172

esp32 sdk does check and return ESP_ERR_INVALID_ARG.

Reply:
Oh yes! That was a bad oversight on my part! I only tested the "WL" pins and forgot to test the regular gpio pins, or it would have failed for sure and I would have caught it.

Reply:
> Also I wonder if we should friendly check the gpio number range and raise badarg if needed as the underlying API doesn't at all unless assertions are enabled. The constant is NUM_BANK0_GPIOS.

Should we add this check to all of the functions, just to be safe?

Reply:
> Should we add this check to all of the functions, just to be safe?

This is what I was suggesting, but I'll be ok if we don't.

Pico sdk's approach to parameters values is to only check them with debug builds. Default and release builds don't check them. It makes sense and we could respect that.

Reply:
I went ahead and added those checks, for the sake of the stability of the VM. At least for the time being, the more unexpected crashes we can rule out the better.